File: backend/src/api/v1/receipts.py
from datetime import date, datetime, time, timezone
from uuid import UUID

from fastapi import APIRouter, Query, status
from src.core.dependencies import (
    AliasRepo,
    CurrentUser,
    Proverkacheka,
    ReceiptRepo,
    TransactionRepo,
    TransactionSvc,
)
from src.schemas.pagination import CursorPage
from src.schemas.receipt import (
    ReceiptCreate,
    ReceiptManualCreate,
    ReceiptParseRequest,
    ReceiptPreviewOut,
    ReceiptResponse,
    ReceiptUpdate,
)
from src.schemas.transaction import TransactionInReceipt, TransactionOut
from src.services.receipts import ReceiptService

router = APIRouter(prefix="/api/v1/receipts", tags=["receipts"])


def _day_bounds(day: date, end_of_day: bool) -> datetime:  # noqa: FBT001
    if end_of_day:
        return datetime.combine(day, time.max, tzinfo=timezone.utc)
    return datetime.combine(day, time.min, tzinfo=timezone.utc)


@router.post("/parse", response_model=ReceiptPreviewOut)
async def parse_receipt(  # noqa: PLR0913
    data: ReceiptParseRequest,
    current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    transaction_service: TransactionSvc,
    alias_repo: AliasRepo,
    proverkacheka: Proverkacheka,
) -> ReceiptPreviewOut:
    normalized, seller_name = await ReceiptService(
        receipt_repo,
        transaction_service,
        alias_repo,
        proverkacheka,
    ).parse(current_user, data)
    return ReceiptPreviewOut.from_normalized(normalized, seller_name)


@router.post(
    "/manual",
    response_model=ReceiptResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_manual_receipt(
    data: ReceiptManualCreate,
    current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    tx_repo: TransactionRepo,
    transaction_service: TransactionSvc,
) -> ReceiptResponse:
    receipt = await ReceiptService(
        receipt_repo,
        transaction_service,
    ).create_manual(current_user, data)
    transactions = await tx_repo.list_by_receipt(receipt.id)
    return ReceiptResponse.from_model(receipt, transactions=transactions)


@router.post("", response_model=ReceiptResponse, status_code=status.HTTP_201_CREATED)
async def create_receipt(  # noqa: PLR0913
    data: ReceiptCreate,
    current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    tx_repo: TransactionRepo,
    transaction_service: TransactionSvc,
    alias_repo: AliasRepo,
    proverkacheka: Proverkacheka,
) -> ReceiptResponse:
    receipt = await ReceiptService(
        receipt_repo,
        transaction_service,
        alias_repo,
        proverkacheka,
    ).create(current_user, data)
    transactions = await tx_repo.list_by_receipt(receipt.id)
    return ReceiptResponse.from_model(receipt, transactions=transactions)


@router.get("", response_model=CursorPage[ReceiptResponse])
async def list_receipts(  # noqa: PLR0913
    _current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    tx_repo: TransactionRepo,
    transaction_service: TransactionSvc,
    limit: int = Query(50, ge=1, le=100),
    cursor: str | None = Query(None),
    date_from: date | None = Query(None),  # noqa: B008
    date_to: date | None = Query(None),  # noqa: B008
    seller: str | None = Query(None, max_length=255),
) -> CursorPage[ReceiptResponse]:
    receipts, next_cursor = await ReceiptService(
        receipt_repo,
        transaction_service,
    ).list_all(
        _current_user,
        limit=limit,
        cursor=cursor,
        date_from=_day_bounds(date_from, end_of_day=False) if date_from else None,
        date_to=_day_bounds(date_to, end_of_day=True) if date_to else None,
        seller=seller,
    )
    # транзакции всех чеков страницы — одним запросом (без N+1)
    receipt_transactions = await tx_repo.list_by_receipts([r.id for r in receipts])
    by_receipt: dict[UUID, list] = {}
    for tx in receipt_transactions:
        if tx.receipt_id is not None:
            by_receipt.setdefault(tx.receipt_id, []).append(tx)
    return CursorPage[ReceiptResponse](
        items=[
            ReceiptResponse.from_model(
                r,
                transactions=by_receipt.get(r.id, []),
            )
            for r in receipts
        ],
        next_cursor=next_cursor,
    )


@router.get("/{receipt_id}", response_model=ReceiptResponse)
async def get_receipt(
    receipt_id: UUID,
    _current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    tx_repo: TransactionRepo,
    transaction_service: TransactionSvc,
) -> ReceiptResponse:
    receipt = await ReceiptService(
        receipt_repo,
        transaction_service,
    ).get(_current_user, receipt_id)
    transactions = await tx_repo.list_by_receipt(receipt.id)
    return ReceiptResponse.from_model(receipt, transactions=transactions)


@router.post(
    "/{receipt_id}/transactions",
    response_model=TransactionOut,
    status_code=status.HTTP_201_CREATED,
)
async def add_receipt_transaction(
    receipt_id: UUID,
    data: TransactionInReceipt,
    _current_user: CurrentUser,
    transaction_service: TransactionSvc,
) -> TransactionOut:
    """Добавляет транзакцию в существующий чек (нумерация продолжается)."""
    tx = await transaction_service.add_to_receipt(
        _current_user,
        receipt_id,
        data,
    )
    return TransactionOut.from_model(tx)


@router.patch("/{receipt_id}", response_model=ReceiptResponse)
async def update_receipt(
    receipt_id: UUID,
    data: ReceiptUpdate,
    _current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    tx_repo: TransactionRepo,
    transaction_service: TransactionSvc,
) -> ReceiptResponse:
    receipt = await ReceiptService(
        receipt_repo,
        transaction_service,
    ).update(_current_user, receipt_id, data)
    transactions = await tx_repo.list_by_receipt(receipt.id)
    return ReceiptResponse.from_model(receipt, transactions=transactions)


@router.delete("/{receipt_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_receipt(
    receipt_id: UUID,
    _current_user: CurrentUser,
    receipt_repo: ReceiptRepo,
    transaction_service: TransactionSvc,
) -> None:
    await ReceiptService(
        receipt_repo,
        transaction_service,
    ).delete(_current_user, receipt_id)

File: frontend/src/api/receipts.ts
/**
 * Receipts domain API — real endpoints, mirrors backend OpenAPI
 * (backend/src/api/v1/receipts.py, schemas/receipt.py).
 *
 * Чек — «коробка» транзакций (receipt_id на каждой транзакции).
 * Транзакции приходят внутри ReceiptResponse.transactions.
 */
import { api } from './client'
import type {
    CursorPage,
    Receipt,
    ReceiptQrDraft,
    ReceiptTransactionDraft,
    ReceiptUpdatePatch,
    Transaction,
} from './types'

export interface ReceiptsPageParams {
    limit?: number
    cursor?: string
    date_from?: string
    date_to?: string
    seller?: string
}

/** GET /api/v1/receipts — cursor-пагинация, новые сверху. */
export function fetchReceiptsPage(params: ReceiptsPageParams = {}): Promise<CursorPage<Receipt>> {
    const search = new URLSearchParams()
    if (params.limit !== undefined) search.set('limit', String(params.limit))
    if (params.cursor) search.set('cursor', params.cursor)
    if (params.date_from) search.set('date_from', params.date_from)
    if (params.date_to) search.set('date_to', params.date_to)
    if (params.seller) search.set('seller', params.seller)
    const qs = search.toString()
    return api.get<CursorPage<Receipt>>(`/api/v1/receipts${qs ? `?${qs}` : ''}`)
}

/** Все чеки (все страницы) — для дашборда. */
export async function fetchAllReceipts(): Promise<Receipt[]> {
    const items: Receipt[] = []
    let cursor: string | undefined
    // защитный лимит: 200 страниц × 100 = 20 000 чеков
    for (let i = 0; i < 200; i++) {
        const page = await fetchReceiptsPage({ limit: 100, cursor })
        items.push(...page.items)
        if (!page.next_cursor) break
        cursor = page.next_cursor
    }
    return items
}

export function getReceipt(id: string): Promise<Receipt> {
    return api.get<Receipt>(`/api/v1/receipts/${id}`)
}

/** GET /api/v1/receipts/{id}/raw — сырые данные чека (raw_json). */
export async function fetchReceiptRaw(id: string): Promise<Record<string, unknown>> {
    return api.get<Record<string, unknown>>(`/api/v1/receipts/${id}/raw`)
}

/** POST /api/v1/receipts — авто-загрузка по QR из proverkacheka (нужен токен). */
export function createReceiptFromQr(draft: ReceiptQrDraft): Promise<Receipt> {
    return api.post<Receipt>('/api/v1/receipts', draft)
}

/** POST /api/v1/receipts/{receipt_id}/transactions — транзакция в существующий чек. */
export function addReceiptTransaction(
    receiptId: string,
    draft: ReceiptTransactionDraft,
): Promise<Transaction> {
    return api.post<Transaction>(`/api/v1/receipts/${receiptId}/transactions`, draft)
}

export function updateReceipt(id: string, patch: ReceiptUpdatePatch): Promise<Receipt> {
    return api.patch<Receipt>(`/api/v1/receipts/${id}`, patch)
}

export function deleteReceipt(id: string): Promise<void> {
    return api.del(`/api/v1/receipts/${id}`)
}
